src/pool-bin/libraries/BinPool.sol

```diff
@@ -200,6 +201,82 @@ library BinPool {
+
+        bytes32 amountOutLeft = swapForY ? amountOut.encodeSecond() : amountOut.encodeFirst();
+        bytes32 amountsIn;
+
+        {
+            uint24 lpFee = params.lpFeeOverride.isOverride()
+                ? params.lpFeeOverride.removeOverrideAndValidate(LPFeeLibrary.TEN_PERCENT_FEE)
+                : slot0Cache.lpFee;
+
+            /// @dev swap fee includes protocolFee (charged first) and lpFee
+            swapState.swapFee = swapState.protocolFee == 0 ? lpFee : swapState.protocolFee.calculateSwapFee(lpFee);
+        }
+
+        /// @notice early return if hook has updated amountIn to 0
+        /// todo: check on this
+        if (amountOut == 0) return (result, swapState);
+
+        while (true) {
+            bytes32 binReserves = self.reserveOfBin[swapState.activeId];
+            if (!binReserves.isEmpty(!swapForY)) {
+                // (bytes32 amountsInWithFees, bytes32 amountsOutOfBin, bytes32 totalFee) =
+                //     binReserves.getAmounts(swapState.swapFee, params.binStep, swapForY, swapState.activeId, amountInLeft);
+                // (bytes amountsOutOfBin, bytes amountsInWithFee, bytes totalFee) =
+                (bytes32 amountsInWithFees, bytes32 amountsOutOfBin, bytes32 totalFee) = binReserves.getAmountsIn(
+                    swapState.swapFee, params.binStep, swapForY, swapState.activeId, amountOutLeft
+                );
+
+                if (amountsInWithFees > 0) {
+                    amountOutLeft = amountOutLeft.sub(amountsOutOfBin);
+                    amountsIn = amountsIn.add(amountsInWithFees);
+
+                    /// @dev calc protocol fee for current bin, totalFee * protocolFee / (protocolFee + lpFee)
+                    bytes32 pFee = totalFee.getExternalFeeAmt(slot0Cache.protocolFee, swapState.swapFee);
+                    if (pFee != 0) {
+                        swapState.feeForProtocol = swapState.feeForProtocol.add(pFee);
+                        amountsInWithFees = amountsInWithFees.sub(pFee);
+                    }
+
+                    self.reserveOfBin[swapState.activeId] = binReserves.add(amountsInWithFees).sub(amountsOutOfBin);
+                }
+            }
+
+            if (amountOutLeft == 0) {
+                break;
+            } else {
+                uint24 nextId = getNextNonEmptyBin(self, swapForY, swapState.activeId);
+                // Equivalent to: if (nextId == 0 || nextId == type(uint24).max) revert BinPool__OutOfLiquidity();
+                assembly ("memory-safe") {
+                    if or(iszero(nextId), eq(nextId, 0xffffff)) {
+                        mstore(0x00, 0x96aa65ad) // Selector BinPool__OutOfLiquidity()
+                        revert(0x1c, 0x04)
+                    }
+                }
+                swapState.activeId = nextId;
+            }
+        }
+
+        if (amountsIn == 0) revert BinPool__InsufficientAmountIn(); // todo
```

Comment:
In what case can this happen ?

out = 1 and price is very low hence in close to 0 ?

Reply:
sorry, let me delete swapExactOut() -- as `swapV2` is the one we should review. 


> out = 1 and price is very low hence in close to 0 ?

yep. in a way more to ensure user get something if their input token !=0